### DISCUSSION TITLE
Prisma.getExtensionContext(this) returning array of functions? (Not the model context)

### DISCUSSION THREAD

**Discussion by outofthisworld** [prisma/prisma#24178] [Q&A]
### Question

I'm creating a plugin, and I was under the impression that when calling `Prisma.getExtensionContext(this)` within a query extension (see below), that I would get a reference to the model which is being acted upon. However, I seem to be getting an array of functions including one named `$alloperations`.

I do have a few other plugins that run before this one, but not sure how this effects things? The docs aren't extensive in this area so would anyone be able to help me out on what is happening?

```ts
  query: {
        $allModels: {
          async findMany<T>(this: T, { query, args, model }) {
            console.log(Prisma.getExtensionContext(model));
            const data = await query(args);
            let result = { data };

            const modelContext = await Prisma.getExtensionContext(this) as T;
            // prints undefined undefined undefined
            console.log(modelContext.name, modelContext.$name, modelContext.scopes);
            return result;
          },
        }
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @outofthisworld 👋 

Thank you for raising this question.

The `findMany` function already receives a `model` argument that holds the actual model instance being acted upon. You can directly use this argument for accessing the model name. Also, you need to do `Prisma.getExtensionContext(this)`, not `Prisma.getExtensionContext(model)`. See this [section](https://www.prisma.io/docs/orm/prisma-client/client-extensions/model#get-the-current-model-name-at-runtime) of the documentation and [here](https://www.prisma.io/docs/orm/prisma-client/client-extensions/model#add-a-custom-method-to-all-models-in-your-schema) as well.

**Reply by outofthisworld** [reply]
Hi @ludralph,

Using `model` when calling `Prisma.getExtensionContext` was just an example - even when I pass `Prisma.getExtensionContext(this)` the result is returning an array of functions, rather the the model being acted upon itself.

for example, when calling it inside the function I get this result:
```
console.log(Prisma.getExtensionContext(this));

// output::
[
  [AsyncFunction: $allOperations],
  [AsyncFunction: $allOperations],
  [AsyncFunction: $allOperations]
]
```

**Reply by ludralph** [reply]
Hi @outofthisworld 

When you use the `query` component, especially with `$allModels`, the context of `this` is different. It doesn't directly represent the model being queried, which is why you were getting an array of `$allOperations` functions instead of the model context. When you use the model component with `$allModels`, the `this` context correctly refers to the current model being operated on. For example:
```js
const prisma = new PrismaClient().$extends({
  model: {
    $allModels: {
      async findMany() {

        const modelContext = await Prisma.getExtensionContext(this);
        
        console.log(modelContext.name, modelContext.$name);
       
      },
    }
  }
});
```
This is why `Prisma.getExtensionContext(this)` returns the expected model context with properties like `name` and `$name`. The [Prisma documentation on model extensions](https://www.prisma.io/docs/orm/prisma-client/client-extensions/model) specifically mentions that you can use Prisma.getExtensionContext(this) to get the current model name at runtime when extending models.

In essence, the `model` extension provides a more direct way to extend and interact with individual models, allowing you to access model-specific context more easily. The `query` extension, on the other hand, is more focused on modifying query behavior across models, which can lead to a different context for `this`.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
